artcommon/artcommonlib/konflux/konflux_db.py

```diff
@@ -36,6 +323,112 @@ def bind(self, record_cls: typing.Type[KonfluxRecord]):
+        coroutine will actually load while others wait for it to complete.
+
+        :param group: Group name (e.g., 'openshift-4.18')
+        """
+        if not self.cache:
+            return
+
+        # Quick check - if already loaded, return immediately
+        if self.cache.is_group_loaded(group):
+            self.logger.debug(f"Cache already loaded for group '{group}'")
+            return
+
+        # Get or create an event for this group (thread-safe)
+        with KonfluxDb._cache_lock:
+            if group not in KonfluxDb._group_loading_events:
+                # Create new event for this group
+                event = asyncio.Event()
+                KonfluxDb._group_loading_events[group] = event
+                should_load = True
+            else:
+                # Another coroutine is already loading
+                event = KonfluxDb._group_loading_events[group]
+                should_load = False
+
+        if not should_load:
+            # Wait for the other coroutine to finish loading
+            self.logger.debug(f"Waiting for another coroutine to load group '{group}'...")
+            await event.wait()
+            self.logger.debug(f"Cache loaded for group '{group}' by another coroutine")
+            return
+
+        # We're the one who will load
+        try:
+            # Double-check after getting permission - might have been loaded while we waited
+            if self.cache.is_group_loaded(group):
+                self.logger.debug(f"Cache already loaded for group '{group}' (loaded while acquiring event)")
+                event.set()  # Signal any waiters
+                return
+
+            self.logger.info(f"Lazy-loading cache for group '{group}' (last {self.cache._cache_days} days)...")
+
+            # Build query for last N days of builds in this group
+            start_time = datetime.now(tz=timezone.utc) - timedelta(days=self.cache._cache_days)
+            where_clauses = [
+                Column('group', String) == group,
+                Column('start_time', DateTime) >= start_time,
+            ]
+
+            order_by_clause = Column('start_time', quote=True).desc()
+
+            # Execute single large query
+            rows = await self.bq_client.select(
+                where_clauses=where_clauses,
+                order_by_clause=order_by_clause,
+                limit=None,  # Get all results
+            )
+
+            # Load all rows into cache (thread-safe operation)
+            builds = [self.from_result_row(row) for row in rows]
+            self.cache.add_builds(builds, group)
+
+            self.logger.info(f"Cache loaded for group '{group}': {len(builds)} builds")
+
+        except Exception as e:
+            self.logger.error(f"Failed to load cache for group '{group}': {e}")
+            raise
+        finally:
+            # Signal completion to any waiting coroutines
+            event.set()
+
+    def cache_stats(self, group: typing.Optional[str] = None) -> dict:
```

Comment:
Currently unused, shall we add a `self.logger.info(self.cache_stats())` in a `__del__` distructor?

Reply:
Just leaving this around for debug. Hopefully we won't need to think about this much after this.